src/core/classes/filebase-storage.ts

```diff
@@ -0,0 +1,282 @@
+import {
+  DuplicateFileNameError,
+  FetchError,
+  UploadError,
+} from "../../common/error";
+import {
+  DEFAULT_IPFS_GATEWAY,
+  PINATA_IPFS_URL,
+  TW_FILEBASE_SERVER_URL,
+  TW_IPFS_SERVER_URL,
+} from "../../constants/urls";
+import { IStorage } from "../interfaces/IStorage";
+import { FileOrBuffer, JsonObject } from "../types";
+import {
+  replaceFilePropertiesWithHashes,
+  replaceHashWithGatewayUrl,
+  resolveGatewayUrl,
+} from "../helpers/storage";
+
+if (!globalThis.FormData) {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  globalThis.FormData = require("form-data");
+}
+
+/**
+ * @internal
+ */
+interface CidWithFileName {
+  // base cid of the directory
+  cid: string;
+
+  // file name of the file without cid
+  fileNames: string[];
+}
+
+/**
+ * IPFS Storage implementation, accepts custom IPFS gateways
+ * @public
+ */
+export class IpfsStorage implements IStorage {
+  private gatewayUrl: string;
+
+  constructor(gatewayUrl: string = DEFAULT_IPFS_GATEWAY) {
```

Comment:
what would be neat here is to pass a network abstraction here that just does the "fetch" and the "upload" - the rest of the logic can be completely shared between different IPFS providers